core/src/main/scala/chisel3/IO.scala

```diff
@@ -44,3 +44,42 @@ object IO {
     iodefClone
   }
 }
+
+private[chisel3] object SecretIO {
+
+  /** Constructs a secret port for the current Module
+    *
+    * This must wrap the datatype used to set the io field of any Module.
+    * i.e. All concrete modules must have defined io in this form:
+    * [lazy] val io[: io type] = IO(...[: io type])
+    *
+    * Items in [] are optional.
+    *
+    * The granted iodef must be a chisel type and not be bound to hardware.
+    *
+    * Also registers a Data as a port, also performing bindings. Cannot be called once ports are
+    * requested (so that all calls to ports will return the same information).
+    * Internal API.
+    */
+  def apply[T <: Data](iodef: => T, module: experimental.BaseModule)(implicit sourceInfo: SourceInfo): T = {
+    val prevId = Builder.idGen.value
+    val data = iodef // evaluate once (passed by name)
+    requireIsChiselType(data, "io type")
+
+    // Clone the IO so we preserve immutability of data types
+    // Note: we don't clone if the data is fresh (to avoid unnecessary clones)
+    val iodefClone =
+      if (!data.mustClone(prevId)) data
+      else
+        try {
+          data.cloneTypeFull
+        } catch {
+          // For now this is going to be just a deprecation so we don't suddenly break everyone's code
+          case e: AutoClonetypeException =>
+            Builder.deprecated(e.getMessage, Some(s"${data.getClass}"))
+            data
+        }
+    module.bindSecretIo(iodefClone)
```

Comment:
I know this is from original IO, but I think this chunk of code is old compatibility mode stuff that is no longer necessary.

Reply:
I think i'll leave it for a future cleanup PR.

Reply:
Nvm, just inlined it.